### PR TITLE
[search] Search quality tests.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: search/search_quality_tests/search_quality_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include "std/cstdio.hpp"
+#include "std/numeric.hpp"
+#include "std/string.hpp"
+#include "std/vector.hpp"
+
+#include "3party/gflags/src/gflags/gflags.h"
+
+using namespace search::tests_support;
+
+#pragma mark Define options
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Do we need this pragma? As I know gFlags, no, we don't.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+{
+  Platform & platform = GetPlatform();
+  return make_unique<storage::CountryInfoReader>(platform.GetReader(PACKED_POLYGONS_FILE),
+                                                 platform.GetReader(COUNTRIES_FILE));
+}
+
+void ReadStringsFromFile(string const & path, vector<string> & result)
+{
+  FILE * f = fopen(path.data(), "r");
+  CHECK(f != nullptr, ("Error when reading strings from", path, ":", string(strerror(errno))));
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: s/f != nullptr/f/

File: search/search_quality_tests/search_quality_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  MY_SCOPE_GUARD(cleanup, [&]
+                 {
+                   fclose(f);
+                 });
+
+  int const kBufSize = 1 << 20;
+  char buf[kBufSize];
+  while (fgets(buf, sizeof(buf), f))
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Please, use iostream library and honest getline() or implement fgets-without-restrictions.

Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Or, if line is wider than 2^20, do exit(-1) with a diagnostic message.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  {
+    string s(buf);
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: 3 copies of buf here :)

File: search/search_quality_tests/search_quality_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    strings::Trim(s);
+    if (!s.empty())
+      result.push_back(s);
+  }
+}
+
+void PrintTopResults(string const & query, vector<search::Result> const & results, size_t n)
+{
+  printf("%s", query.data());
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Could you please add a comment how the output looks like?

File: search/search_quality_tests/search_quality_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    platform.SetResourceDir(FLAGS_data_path);
+
+  if (!FLAGS_mwm_path.empty())
+    platform.SetWritableDirForTests(FLAGS_mwm_path);
+
+  LOG(LINFO, ("writable dir =", platform.WritableDir()));
+  LOG(LINFO, ("resources dir =", platform.ResourcesDir()));
+
+  classificator::Load();
+  auto infoGetter = CreateCountryInfoGetter();
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Please, inline country info getter instantiation to TestSearchEngine ctor call.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  for (auto const & mwm : mwms)
+  {
+    LOG(LINFO, ("Registering map:", mwm.GetCountryName()));
+    engine.RegisterMap(mwm);
+  }
+
+  vector<string> queries;
+  string queriesPath = FLAGS_queries_path;
+  if (queriesPath.empty())
+    queriesPath = platform.WritableDir() + kDefaultQueriesPathSuffix;
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: JoinFoldersToPath please.

File: search/search_quality_tests/search_quality_tests.cpp (hunk 1 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 2 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 6 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 7 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 3 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 4 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 5 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

**Comment by mpimenov** [comment]
PTAL

File: search/search_quality_tests/search_quality_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  {
+    return make_unique<search::v2::SearchQueryV2>(index, categories, suggests, infoGetter);
+  }
+};
+
+void ReadStringsFromFile(string const & path, vector<string> & result)
+{
+  ifstream stream(path.c_str());
+  // Using errno here gives no guarantees but is better than nothing.
+  CHECK(stream.is_open(), ("Error when reading strings from", path, ":", string(strerror(errno))));
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Just write
```c++
CHECK(..., ("Can't open", path));
```
This magic code with errno does not work.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  MY_SCOPE_GUARD(cleanup, [&]
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: stream is automatically closed when destructed.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+{
+  cout << query;
+  for (size_t i = 0; i < n; ++i)
+  {
+    if (n > 1)
+      cout << "\n";
+    cout << "\t";
+    if (i < results.size())
+      // todo(@m) Print more information: coordinates, viewport, etc.
+      cout << string(results[i].GetString());
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: No need in explicit const char * -> string conversion, emit result strings as is.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    else
+      cout << kEmptyResult;
+  }
+  cout << "\n";
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: endl

File: search/search_quality_tests/search_quality_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+// If n == 1, prints the query and the top result separated by a tab.
+// Otherwise, prints the query on a separate line
+// and then prints n top results on n lines starting with tabs.
+void PrintTopResults(string const & query, vector<search::Result> const & results, size_t n)
+{
+  cout << query;
+  for (size_t i = 0; i < n; ++i)
+  {
+    if (n > 1)
+      cout << "\n";
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: endl, please.

File: search/search_quality_tests/search_quality_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+  vector<string> queries;
+  string queriesPath = FLAGS_queries_path;
+  if (queriesPath.empty())
+    queriesPath = my::JoinFoldersToPath(platform.WritableDir(), kDefaultQueriesPathSuffix);
+  ReadStringsFromFile(queriesPath, queries);
+
+  for (string const & query : queries)
+  {
+    // todo(@m) Viewport and position should belong to the query info.
```
Comment (ygorshenin) [inline search/search_quality_tests/search_quality_tests.cpp]: Probably it's worth to pass them through command line. It's up to you whether it worth to implement it right now or delay until we will need custom viewport and position.

File: search/search_tests_support/test_search_engine.hpp [hunk 14]
```diff
@@ -29,6 +29,8 @@ class TestSearchEngine : public Index
   TestSearchEngine(string const & locale, unique_ptr<storage::CountryInfoGetter> && infoGetter);
   TestSearchEngine(string const & locale, unique_ptr<storage::CountryInfoGetter> && infoGetter,
                    unique_ptr<search::SearchQueryFactory> && factory);
+  TestSearchEngine(string const & locale, unique_ptr<::search::SearchQueryFactory> && factory);
```
Comment (ygorshenin) [inline search/search_tests_support/test_search_engine.hpp]: Pass unique ptr by value, not by rvalue-reference.
Also, feel free to fix other constructors in this file.

File: search/search_quality_tests/search_quality_tests.cpp (hunk 8 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: Strangely enough, it does.
But ok, removed.

File: search/search_quality_tests/search_quality_tests.cpp (hunk 9 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 10 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 11 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: endl flushes too much

File: search/search_quality_tests/search_quality_tests.cpp (hunk 13 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: No. Different queries having different viewports is a completely valid case.

File: search/search_tests_support/test_search_engine.hpp (hunk 14 above)
Comment (mpimenov) [inline search/search_tests_support/test_search_engine.hpp]: done

File: search/search_quality_tests/search_quality_tests.cpp (hunk 12 above)
Comment (mpimenov) [inline search/search_quality_tests/search_quality_tests.cpp]: done

**Comment by mpimenov** [comment]
PTAL

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**